### PR TITLE
Delete car endpoints

### PR DESCRIPTION
**Pull request by RaphealOlumofe**

### CODE REVIEW

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (IDWalsh) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: DOn't need to initialise a new list here. What does it even do?

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (IDWalsh) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: This isn't compatible with the update single car method. E.g. I can set some other index as the car id in the other one and add it with any key I like. Not with the order index.

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (IDWalsh) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: Don't iterate maps. Just to map.getKey(car.getId)

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: This line, at the moment, is redundant. You are putting the car into the map with the key `map.size()+1` which works but isn't an ideal way to express the relationship between the key and the value.  Secondly, you are then setting the ID of the car that's passed in via the parameter (doesn't it need an ID anyways in the request?) after you put it in the map. So this car with the new id is never used.

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: Lets avoid iterating through the whole map. If our map is 1000 cars long and our request is only 1 car, then we should only need to check the 1 car. We are not utilising the key feature of maps.

https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#getOrDefault-java.lang.Object-V-

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: Not bad but a bit hard to understand. Can you explain `e -> carId`? I think there is an easier way to do this.

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: This list doesn't need to be here. You now have two equal variables in the same scope. `carList` only exists in the method as does `currentCarList` and they are equivalent (until you update/change one or the other)

**Review by timlarochebbc — CHANGES_REQUESTED** [review]

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (IDWalsh) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: Maps also have a delete key method

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (RaphealOlumofe) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: resolved

Comment (IDWalsh) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: You don't need to reply resolved ;-) 
Also it's not really resolved until someone else has seen it and resolved it.

Comment (RaphealOlumofe) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: kool

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (RaphealOlumofe) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: thanks

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java:21]: We could simply do `return map.get(carId)`

Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: We could simply do `return map.get(carId)` as we don't really do anything to the object we get back

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java:31]: Nice!

**Review by timlarochebbc — COMMENTED** [review]
Looks good, just need to sort out the merge conflict before approval

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java:19]: This endpoint doesn't work. We need to parse the list pathVariable somehow or include the list of ids in the request body instead of in the path. At the moment we can't get a `List<Integer>` from `{list}` (FWIW, you can but it will always be size 1)

**Review by timlarochebbc — CHANGES_REQUESTED** [review]
Almost there I think. The endpoint names are a little hard to follow along with. Perhaps change them or group them up in the code to have all the GETs together and all the POSTs together.

Comment about `getBatchCars` at `/cars/{list}` not working.

Also bring your attention to this nitpick that I can't seem to comment on
```
    @PutMapping(path = "car", consumes = {"application/json"})
    public Map<Integer, Car> updateCars(@RequestBody Car car) {
        Car entry = map.get(car.getId());
        entry = car;
        map.put(entry.getId(), entry);
        return map;
    }

    @PutMapping(path = "cars")
    public Map<Integer, Car> batchUpdateCars(@RequestBody Map<Integer, Car> m) {

        for (Map.Entry<Integer, Car> entry : m.entrySet()) {
            int key = entry.getKey();
            Car value = entry.getValue();
            Car updatedCar = map.get(key);
            updatedCar = value;
            map.put(key, updatedCar);
        }
        return map;
    }
```

We have some unnecessary variables above. 
Consider
```
        Car entry = map.get(car.getId());
        entry = car;
        map.put(entry.getId(), entry);
        return map;
```
this can be simplified to 
```
        map.put(car.getId(), car);
        return map;
```
the initial value you set for entry is never used so no point in assigning it.

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (petermcneil) [reply frontend/src/main/java/starter/project/frontend/CarController.java:19]: GET requests shouldn't have a request body.
This works in spring magically. The request should look like `/cars/1,2,3,4`

Comment (timlarochebbc) [reply frontend/src/main/java/starter/project/frontend/CarController.java:19]: TIL that's a thing. Apologies, feel free to mark this as resolved

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: This was marked resolved but not changed? Is there justification to keeping the variable

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (timlarochebbc) [inline frontend/src/main/java/starter/project/frontend/CarController.java:45]: One last change, in `batchUpdateCars` we also have the same problem as before with the unused variable. A similar change like you did here would be good.

**Review by timlarochebbc — CHANGES_REQUESTED** [review]
A lot neater with the grouping now. Left a comment about the other PUT method. Also there was a comment marked resolved but not updated.

File: frontend/src/main/java/starter/project/frontend/CarController.java
Comment (stuartatwork) [inline frontend/src/main/java/starter/project/frontend/CarController.java]: I think, instead of looping through the map, you could just do:

`map.putAll(m);`

Comment (RaphealOlumofe) [reply frontend/src/main/java/starter/project/frontend/CarController.java]: thanks

**Comment by stuartatwork** [comment]
This looks good. Only thing outstanding that I can see is Tim and Iain's comment about a simpler implementation of the delete key. Have you had a look at the Map's javadocs? 

https://docs.oracle.com/javase/8/docs/api/java/util/Map.html

There is a `remove` method in there which should achieve what you want to do here.

I am pretty sure that is the last change needed.

**Comment by stuartatwork** [comment]
That would work. But you don't need to call the keyset and remove it from that set. You can just call the remove on the map itself.

**Review by stuartatwork — APPROVED** [review]

**Pull request merged**
